esphome/components/ina2xx_base/ina2xx_base.h

```diff
@@ -0,0 +1,245 @@
+#pragma once
+
+#include "esphome/core/component.h"
+#include "esphome/components/sensor/sensor.h"
+
+namespace esphome {
+namespace ina2xx_base {
+
+enum RegisterMap : uint8_t {
+  REG_CONFIG = 0x00,
+  REG_ADC_CONFIG = 0x01,
+  REG_SHUNT_CAL = 0x02,
+  REG_SHUNT_TEMPCO = 0x03,
+  REG_VSHUNT = 0x04,
+  REG_VBUS = 0x05,
+  REG_DIETEMP = 0x06,
+  REG_CURRENT = 0x07,
+  REG_POWER = 0x08,
+  REG_ENERGY = 0x09,
+  REG_CHARGE = 0x0A,
+  REG_DIAG_ALRT = 0x0B,
+  REG_SOVL = 0x0C,
+  REG_SUVL = 0x0D,
+  REG_BOVL = 0x0E,
+  REG_BUVL = 0x0F,
+  REG_TEMP_LIMIT = 0x10,
+  REG_PWR_LIMIT = 0x11,
+  REG_MANUFACTURER_ID = 0x3E,
+  REG_DEVICE_ID = 0x3F
+};
+
+enum AdcRange : uint16_t {
+  ADC_RANGE_0 = 0,
+  ADC_RANGE_1 = 1,
+};
+
+enum AdcTime : uint16_t {
+  ADC_TIME_50US = 0,
+  ADC_TIME_84US = 1,
+  ADC_TIME_150US = 2,
+  ADC_TIME_280US = 3,
+  ADC_TIME_540US = 4,
+  ADC_TIME_1052US = 5,
+  ADC_TIME_2074US = 6,
+  ADC_TIME_4120US = 7,
+};
+
+enum AdcAvgSamples : uint16_t {
+  ADC_AVG_SAMPLES_1 = 0,
+  ADC_AVG_SAMPLES_4 = 1,
+  ADC_AVG_SAMPLES_16 = 2,
+  ADC_AVG_SAMPLES_64 = 3,
+  ADC_AVG_SAMPLES_128 = 4,
+  ADC_AVG_SAMPLES_256 = 5,
+  ADC_AVG_SAMPLES_512 = 6,
+  ADC_AVG_SAMPLES_1024 = 7,
+};
+
+union ConfigurationRegister {
+  uint16_t raw_u16;
+  struct {
+    uint16_t reserved_0_3 : 4;  // Reserved
+    AdcRange ADCRANGE : 1;      // Shunt measurement range 0: ±163.84 mV, 1: ±40.96 mV
+    bool TEMPCOMP : 1;          // Temperature compensation enable
+    uint16_t CONVDLY : 8;       // Sets the Delay for initial ADC conversion in steps of 2 ms.
+    bool RSTACC : 1;            // Reset counters
+    bool RST : 1;               // Full device reset
+  } __attribute__((packed));
+};
+
+union AdcConfigurationRegister {
+  uint16_t raw_u16;
+  struct {
+    AdcAvgSamples AVG : 3;
+    AdcTime VTCT : 3;    // Voltage conversion time
+    AdcTime VSHCT : 3;   // Shunt voltage conversion time
+    AdcTime VBUSCT : 3;  // Bus voltage conversion time
+    uint16_t MODE : 4;
+  } __attribute__((packed));
+};
+
+union TempCompensationRegister {
+  uint16_t raw_u16;
+  struct {
+    uint16_t TEMPCO : 14;
+    uint16_t reserved : 2;
+  } __attribute__((packed));
+};
+
+union DiagnosticRegister {
+  uint16_t raw_u16;
+  struct {
+    bool MEMSTAT : 1;
+    bool CNVRF : 1;
+    bool POL : 1;
+    bool BUSUL : 1;
+    bool BUSOL : 1;
+    bool SHNTUL : 1;
+    bool SHNTOL : 1;
+    bool TMPOL : 1;
+    bool RESERVED1 : 1;
+    bool MATHOF : 1;
+    bool CHARGEOF : 1;
+    bool ENERGYOF : 1;
+    bool APOL : 1;
+    bool SLOWALERT : 1;
+    bool CNVR : 1;
+    bool ALATCH : 1;
+  } __attribute__((packed));
+};
+
+class INA2XX : public PollingComponent {
+ public:
+  void setup() override;
+  float get_setup_priority() const override;
+  void update() override;
+  void loop() override;
+  void dump_config() override;
+
+  void set_shunt_resistance_ohm(float shunt_resistance_ohm) { shunt_resistance_ohm_ = shunt_resistance_ohm; }
```

Comment:
> All uses of class members and member functions should be prefixed with this-> to distinguish them from global functions in code review.

https://esphome.io/guides/contributing#codebase-standards

```suggestion
  void set_shunt_resistance_ohm(float shunt_resistance_ohm) { this->shunt_resistance_ohm_ = shunt_resistance_ohm; }
```